### DISCUSSION TITLE
Use in Streamlit ?

### DISCUSSION THREAD

**Discussion by Elmetayer** [developmentseed/lonboard#342] [Q&A] [answered]
I was wondering if it is possible to use the library out of notebooks and in "standalone" applications, such as for example Streamlit ?

**Comment by kylebarron** [comment, marked as answer]
I'm not familiar with Streamlit, so I can't give any direct advice.

In general, you can use [`to_html`](https://developmentseed.org/lonboard/latest/api/map/#lonboard.Map.to_html) to export a map as static HTML. Maybe that will help?

**Comment by cboettig** [comment]
Thanks, just chiming in here that I'd love to see this as well.  Streamlit is an extremely popular simple python framework for reactive apps and a first-class citizen on [huggingface spaces](https://huggingface.co/docs/hub/en/spaces-overview), and streamlit already uses deck.gl / pydeck as the default mapping engine for it's [map](https://docs.streamlit.io/develop/api-reference/charts/st.map).

I think we can get pretty far with to_html(), but it looks like there is already at least third party support for **bi-directional** communication, e.g. in [streamlit-deckgl](https://pypi.org/project/streamlit-deckgl/).  Bi-directional support opens up a lot of possibilities for a web-based app, e.g. compare to the bi-directional input on [folium maps for streamlit](https://folium.streamlit.app/).  This lets us write reactive python apps that can respond to user interactions with the map such as the drawing tools, etc.

**Reply by kylebarron** [reply]
> third party support for **bi-directional** communication, e.g. in [streamlit-deckgl](https://pypi.org/project/streamlit-deckgl/)

I took a quick look at its source (it's not clear if it's hosted on github anywhere), and it looks like it injects its own custom HTML and JS into the streamlit component. I'd say this is out of scope for lonboard itself, but I'd be supportive of anyone else who wants to write this connector.

> Bi-directional support opens up a lot of possibilities for a web-based app

I do agree, but it would be simpler if streamlit offered first-class support for jupyter widgets. E.g. I have a [specific example to integrate with shiny](https://developmentseed.org/lonboard/latest/ecosystem/shiny/), which is in-scope because shiny supports widgets directly (even reactively!), so I don't have to write a custom connector.

**Comment by kiasari** [comment]
@cboettig Could you find a way to display the Map object in Streamlit?

**Comment by relativityhd** [comment]
Hi all,
I have created (or better let create) a third party streamlit connector for lonboard, it is now published on PyPi.

[Pypi](https://pypi.org/project/streamlit-lonboard/)
[GitHub](https://github.com/relativityhd/streamlit-lonboard)

It is mostly AI generated with Claude Code (5th gen models), but I have reviewed the code and it looks okay-ish to me. I have tested it on one of my projects, and it works well enough. Of course, first-party support either by lonboard or by streamlit by supporting widgets would be better, but I really needed the nice performance of lonboard in my dashboards. I hadn't the time to build the connector by myself, so I used this project to test how far I can get with Claude.

I am open to feedback, I hope this may help someone!

@kylebarron, maybe this could be added to the [ecosystem page of your docs](https://developmentseed.org/lonboard/latest/ecosystem/)?

**Reply by kylebarron** [reply]
Hi, it's nice that you're working on this; it's a hard problem.

I took a quick look into how streamlit-lonboard is set up, and it looks like it's fully reimplementing the frontend of Lonboard. It's relying on non-public implementation details of how Lonboard currently manages serialization of layers (or actually completely changes the serialization I guess, since it sends Arrow IPC instead of Parquet).

It looks like you're only using the Python models as a serialization source, but fully reimplementing the frontend.

This doesn't feel like stable architectural footing in my opinion, and I'd be hesitant to recommend it as a lonboard connector, because it's actually a from-scratch reimplementation using deck.gl. (This is why I never did any further work with streamlit, because it looked like I would have to reimplement the whole library to get it to work with their tooling)

**Reply by relativityhd** [reply]
Fair enough, I agree that the reimplementation of the frontend breaks with a lot of your stuff. On the other hand, as far as I understood this is what Streamlit expects from third party components... What would be a better, more stable architecture for such a connector in your opinion?

As mentioned, the way it is currently implemented is really a monkey-found-AI solution, as I do not have the time to create a quality solution for that on my own. I know it's not good but for my use case it works well enough.

**Reply by kylebarron** [reply]
I don't think there is a better, more stable architecture that we can implement ourselves. I think we're locked in to what Streamlit provides, and they don't provide any reasonable interface for working with Jupyter widgets.

The fix for this is for them to provide native first-class support for jupyter widgets.
